test/topology/test_cluster_features.py

```diff
@@ -0,0 +1,163 @@
+TEST_FEATURE_ENABLE_ERROR_INJECTION = "features_enable_test_feature"
+ERROR_INJECTIONS_AT_STARTUP_CONFIG_KEY = "error_injections_at_startup"
+
+
+async def get_error_injections_enabled_at_startup(manager: ManagerClient, srv: ServerInfo) -> Set[str]:
+    # TODO: An "error injection enabled at startup" might not be a string, but a dictionary
+    #       with some options. The tests in this module only use strings, but it's worth
+    #       keeping that in mind here in case dicts start being used.
+    config = await manager.server_get_config(srv.server_id)
+    injections = config.get(ERROR_INJECTIONS_AT_STARTUP_CONFIG_KEY) or []
+    assert isinstance(injections, list)
+    assert all(isinstance(inj, str) for inj in injections)
+    return set(injections)
+
+
+async def turn_on_test_feature_and_restart(manager: ManagerClient, srv: ServerInfo) -> None:
+    logging.info(f"Reconfiguring and restarting node {srv} to support {TEST_FEATURE_NAME}")
+    await manager.server_stop_gracefully(srv.server_id)
+    injections = await get_error_injections_enabled_at_startup(manager, srv)
+    injections.add(TEST_FEATURE_ENABLE_ERROR_INJECTION)
+    await manager.server_update_config(srv.server_id, ERROR_INJECTIONS_AT_STARTUP_CONFIG_KEY, list(injections))
+    await manager.server_start(srv.server_id)
+
+
+async def turn_off_test_feature_and_restart(manager: ManagerClient, srv: ServerInfo, expected_error: Optional[str] = None) -> None:
+    logging.info(f"Reconfiguring and restarting node {srv} to disable support of {TEST_FEATURE_NAME}")
+    await manager.server_stop_gracefully(srv.server_id)
+    injections = await get_error_injections_enabled_at_startup(manager, srv)
+    injections.remove(TEST_FEATURE_ENABLE_ERROR_INJECTION)
+    await manager.server_update_config(srv.server_id, ERROR_INJECTIONS_AT_STARTUP_CONFIG_KEY, list(injections))
+    await manager.server_start(srv.server_id, expected_error)
+
+
+@pytest.mark.asyncio
+async def test_cluster_upgrade(manager: ManagerClient) -> None:
+    """Simulates an upgrade of a cluster by doing a rolling restart
+       and marking the test-only feature as supported on restarted nodes.
+    """
+    servers = await manager.running_servers()
+
+    cql = manager.cql
+    hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+
+    for srv in servers:
+        host = (await wait_for_cql_and_get_hosts(cql, [srv], time.time() + 60))[0]
+
+        # The feature should not be advertised as supported until we enable it
+        # via error injection.
+        assert TEST_FEATURE_NAME not in await get_supported_features(cql, host)
+
+        # Until all nodes are updated to support the test feature, that feature
+        # should not be considered enabled by any node.
+        for host in hosts:
+            assert TEST_FEATURE_NAME not in await get_enabled_features(cql, host)
+
+        await turn_on_test_feature_and_restart(manager, srv)
+        cql = await reconnect_driver(manager)
```

Comment:
I don't understand why `reconnect_driver` is necessary; doesn't `wait_for_cql_and_get_hosts` do the job already?

Reply:
I might have misattributed some failure that I had when running the tests locally to scylladb/python-driver#170 and applied the `reconnect_driver` workaround. My installed version of the driver should not have this bug, and the tests don't seem to be flaky when I run them, so I'll drop the use of `reconnect_driver`.